src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -484,6 +484,44 @@ namespace shogun
 			    ->add_col_vec(A, i, b, result, alpha, beta);
 		}
 
+		/** Calculates pseudo inverse A+ from the singular value
+		 * decomposition A = UDV^T , by  A^+ = V(D+)U^T. A+ is unique and does
+		 * always exist.
+		 * User should pass an appropriately pre-allocated memory matrix
+		 * or pass the operand argument A as a result.
+		 *
+		 * @param A The matrix
+		 * @param result The matrix that saves the result
+		 */
+		template <typename T>
+		void pinv_self_adjoint(const SGMatrix<T>& A, SGMatrix<T>& result)
+		{
+
+			REQUIRE(
+			    result.num_rows == A.num_rows && result.num_cols == A.num_cols,
```

Comment:
pls check how scipy/numpy pinv behave
They do work for non symmetric matrices.
Best thing is you do unit tests

* symmetric psd matrix, using self_adjoint=true
* square matrix that is not PSD, using self_adjoint=false
* non-square matrix (more rows that cols), using self_adjoint=false
* non-square matrix (more cols than rows), using self_adjoint=false

I suggest using some 3x3 matrix, and then do a unit test that compares against pinv in numpy

Reply:
ah and also tests that assert for raised errors when self-adjoint=true but a non-square matrix is passed